third_party/move/tools/move-smith/src/ast.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! An abstract syntax tree for the Move language used by the MoveSmith fuzzer.
+//! The AST is taken mostly from `third_party/move/move-compiler/src/parser/ast.rs`.
+//! Ideally when the fuzzer becomes more mature, this AST will converge to the
+//! parser's AST and we might be able to reuse the parser's AST directly.
+
+use crate::{names::Identifier, types::Type};
+use arbitrary::Arbitrary;
+use num_bigint::BigUint;
+
+/// The collection of modules and scripts that make up a Move program.
+/// This is the final output of the MoveSmith fuzzer.
+/// This should be runnable as a transactional test.
+#[derive(Debug, Clone)]
+pub struct CompileUnit {
+    pub modules: Vec<Module>,
+    pub scripts: Vec<Script>,
+}
+
+/// A Move module.
+#[derive(Debug, Clone)]
+pub struct Module {
+    // pub attributes: Vec<Attributes>,
+    // pub address: Option<LeadingNameAccess>,
+    pub name: Identifier,
+    // pub is_spec_module: bool,
+    pub functions: Vec<Function>,
+    pub structs: Vec<StructDefinition>,
+    // pub uses: Vec<UseDecl>,
+    // pub friends: Vec<FriendDecl>,
+    // pub constants: Vec<Constant>,
+    // pub specs: Vec<SpecBlock>,
+}
+
+/// A simplified Move Script.
+/// The script only contains a `main` function.
+/// The `main` function only consists of a sequence of function calls.
+#[derive(Debug, Clone)]
+pub struct Script {
+    // pub uses: Vec<UseDecl>,
+    pub main: Vec<FunctionCall>,
+}
+
+/// A function definition.
+/// The return statement is separated from the body to simplify verifying the
+/// generated function has a valid return.
+#[derive(Debug, Clone)]
+pub struct Function {
+    // pub attributes: Vec<Attributes>,
+    pub visibility: Visibility,
+    pub signature: FunctionSignature,
+    /// `None` indicates no specifiers given, `Some([])` indicates the `pure` keyword has been
+    /// used.
+    // pub access_specifiers: Option<Vec<AccessSpecifier>>,
+    pub name: Identifier,
+    // pub inline: bool,
+    pub body: Option<FunctionBody>,
+
+    pub return_stmt: Option<Expression>,
```

Comment:
I am assuming we are limiting to at most one return statement, only to get started with and not as limitation.

Reply:
Good catch! Yes currently only one return will be generated but we could easily generate more returns later. Will keep this point in mind but will add the functionality in future PRs.